miniapps/shifted/integ_algoim.hpp

```diff
@@ -0,0 +1,351 @@
+#ifndef INTEG_ALGOIM_HPP
+#define INTEG_ALGOIM_HPP
+
+#include <mfem.hpp>
+
+#ifdef MFEM_USE_ALGOIM
+#include <algoim_quad.hpp>
+#endif
+
+
+namespace mfem
+{
+
+
+//define templated element bases
+namespace TmplPoly_1D
+{
+
+/// Templated version of CalcBinomTerms
+template<typename float_type>
+void CalcBinomTerms(const int p, const float_type x, const float_type y,
+                    float_type* u)
+{
+   if (p == 0)
+   {
+      u[0] = float_type(1.);
+   }
+   else
+   {
+      int i;
+      const int *b = Poly_1D::Binom(p);
+      float_type z = x;
+      for (i = 1; i < p; i++)
+      {
+         u[i] = b[i]*z;
+         z *= x;
+      }
+      u[p] = z;
+      z = y;
+      for (i--; i > 0; i--)
+      {
+         u[i] *= z;
+         z *= y;
+      }
+      u[0] = z;
+   }
+}
+
+/// Templated version of CalcBinomTerms
+template<typename float_type>
+void CalcBinomTerms(const int p, const float_type x, const float_type y,
+                    float_type* u, float_type* d)
+{
+   if (p == 0)
+   {
+      u[0] = float_type(1.);
+      d[0] = float_type(0.);
+   }
+   else
+   {
+      int i;
+      const int *b = Poly_1D::Binom(p);
+      const float_type xpy = x + y, ptx = p*x;
+      float_type z = float_type(1.);
+
+      for (i = 1; i < p; i++)
+      {
+         d[i] = b[i]*z*(i*xpy - ptx);
+         z *= x;
+         u[i] = b[i]*z;
+      }
+      d[p] = p*z;
+      u[p] = z*x;
+      z = float_type(1.);
+      for (i--; i > 0; i--)
+      {
+         d[i] *= z;
+         z *= y;
+         u[i] *= z;
+      }
+      d[0] = -p*z;
+      u[0] = z*y;
+   }
+
+}
+
+/// Templated version of CalcDBinomTerms
+template <typename float_type>
+void CalcDBinomTerms(const int p, const float_type x,
+                     const float_type y, float_type* d)
+{
+   if (p == 0)
+   {
+      d[0] = 0.;
+   }
+   else
+   {
+      int i;
+      const int *b = Poly_1D::Binom(p);
+      const float_type xpy = x + y, ptx = p*x;
+      float_type z = float_type(1.);
+
+      for (i = 1; i < p; i++)
+      {
+         d[i] = b[i]*z*(i*xpy - ptx);
+         z *= x;
+      }
+      d[p] = p*z;
+      z = float_type(1.);
+      for (i--; i > 0; i--)
+      {
+         d[i] *= z;
+         z *= y;
+      }
+      d[0] = -p*z;
+   }
+}
+
+/// Templated evaluation of Bernstein basis
+template <typename float_type>
+void CalcBernstein(const int p, const float_type x, float_type *u)
+{
+   CalcBinomTerms(p, x, 1. - x, u);
+}
+
+
+/// Templated evaluation of Bernstein basis
+template <typename float_type>
+void CalcBernstein(const int p, const float_type x,
+                   float_type *u, float_type *d)
+{
+   CalcBinomTerms(p, x, 1. - x, u, d);
```

Comment:
Similar to my previous comment, the CalcBernstein functions can also be modified in a similar manner.

Reply:
These templates reproduce exactly the code in the non-templated mfem implementation, therefore I would like to keep them as they are.